Add dependencies to allow using additional logstash-logback appenders in logback.xml

sunix opened the pull request:
### What does this PR do?
- Add dependencies to allow using additional logstash-logback appenders in logback.xml
- Add a new `include` entry in logback.xml to allow adding more appenders from a custom assembly

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/6537

### New behavior
Che users can now include logstash appenders in their custom assemblies for instance in ` `/home/user/eclipse-che-5.19.0-SNAPSHOT/tomcat/conf/logback-additional-appenders.xml`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<included>
    <appender name="file-json" class="ch.qos.logback.core.rolling.RollingFileAppender">
        <file>${che.logs.dir}/logs/catalina.log.json</file>
        <filter class="ch.qos.logback.classic.filter.ThresholdFilter">
          <level>info</level>
        </filter>
        <rollingPolicy class="ch.qos.logback.core.rolling.TimeBasedRollingPolicy">
            <fileNamePattern>${che.logs.dir}/archive/%d{yyyy/MM/dd}/catalina.log.json</fileNamePattern>
            <maxHistory>${max.retention.days}</maxHistory>
        </rollingPolicy>
        <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
    </appender>

    <root level="${che.logs.level:-INFO}">
        <appender-ref ref="file-json"/>
    </root>
</included>
```

### Tests written?
No

### Docs updated?
TODO after feedbacks on implementation

skabashnyuk commented:
How Eclipse Che user will benifit from this?

vparfonov commented:
I don't see any profit from this. I think if OSIO need custom logger it should be solved ob OSIO side

skabashnyuk commented:
```
 <include optional="true" file="${che.local.conf.dir}/logback/logback-additional-appenders.xml"/>
```
is made just for the case if some one needs custom logger. 
@sunix can you configre json appender in your custom ```logback-additional-appenders.xml```

sunix commented:
@skabashnyuk
> How Eclipse Che user will benefit from this?
Eclipse Che 'user' will have the json file appender for logstash just by setting an var environment.
It would be also easy to add an appender that sends event directly to logstash with the added libraries.

> @sunix can you configre json appender in your custom logback-additional-appenders.xml
We can consider using logback-additional-appenders.xml but it's located in ${che.local.conf.dir} which doesn't make it easy to package as the location can be setted elsewhere. I would rather add a new line ```
<include optional="true" file="${catalina.home}/conf/logback-additional-appenders.xml"/>
```
and remove janino and the conditional if in logback.xml if the env variable is not prefered.

> Do we have a CQ for janino?
If we decide to keep this implementation I will create the CQs for jamino and other new dependencies.

> What kind of class-loading issues have you met? What was the problem and how it was solved?
The problem was that some class from logback are loaded by the classloader of the webapp which is by default higher than the classloader of the the tomcat/lib, and other.
So when using logstash-logback or janino we get the errors: https://gist.github.com/sunix/29177cb1afc467752763598cfc419502

sunix commented:
> I don't see any profit from this. I think if OSIO need custom logger it should be solved ob OSIO side

@vparfonov I don't think that osio is the only project that would get benefits of it. Nowadays json logs for logstash/ELK or any other event based system is very widely used.

skabashnyuk commented:
@sunix I've looked on class loading error you have. Looks like it's related to ```janino``` location. If we would not use it(janino) - will we have the same issue with class loading?

sunix commented:
@skabashnyuk without janino:

```
08:29:35,623 |-ERROR in ch.qos.logback.core.joran.util.PropertySetter@4277f097 - A "net.logstash.logback.encoder.LogstashEncoder" object is not assignable to a "ch.qos.logback.core.encoder.Encoder" variable.
```

https://gist.github.com/sunix/29177cb1afc467752763598cfc419502#file-logstashencoder-catalina-out

skabashnyuk commented:
@sunix I think I know the cause of the class loading issue. 
Actually, we have two instances of logback logger. One is for tomcat itself. It classes is loaded from tomcat-juli.jar Another is war based, because of that, we need to put logback+slf4j jars in war too https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/pom.xml#L265-L276  Maybe it would be reasonable to put same jars in war too and it will solve class loading issue. 
I think it would be better if we stay in standard J2EE class loading model, because some library in future may expect that. WDYT?

BTW what is the reason to add ```ch.qos.logback:logback-classic```?

sunix commented:
@skabashnyuk I will give a try. For information, `delegate=true` is by default in some J2EE application containers (Glassfish ...)  but not in others.
I added all the dependencies that were needed for `logstash-logback-encoder`, https://github.com/logstash/logstash-logback-encoder seems that it's needed for LoggingEvent. I've added all the dependencies the the library requires so user could also configure other available logstash-logback-encoder appenders and not just the file json output.

By the way, I'm going to update this pull request and not use jamino nor the che.env property: I will add another <includes> in logback.xml that would refers to a file in tomcat/conf. Would be easier to provide additional appenders from assemblies without having to patch logback.xml.

sunix commented:
It works well by adding `logstash-logback-encoder`, `logback-core` and `logback-access` to wsmaster pom.xml dependencies.
Now, it's not an issue for me but we have these logs:

```
2017-09-29 13:24:32,384[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 923]        - Deploying web application archive /assembly/tomcat/webapps/wsmaster.war
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/assembly/tomcat/lib/logback-classic-1.2.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/assembly/tomcat/webapps/wsmaster/WEB-INF/lib/logback-classic-1.2.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [ch.qos.logback.classic.util.ContextSelectorStaticBinder]
```

sunix commented:
@skabashnyuk @riuvshin @vparfonov @benoitf is this one ok to be merged ? ( once https://github.com/eclipse/che-dependencies/pull/60 is ok to be merged )

skabashnyuk commented:
@sunix  can you measure how bigger tomcat become with new libs?

sunix commented:
@skabashnyuk the tomcat itself,
with new libs:

    -rw-r--r--. 1 user user  12M Nov  7 14:07 tomcat-juli.zip

before:

    -rw-r--r--. 1 user user 8.9M Nov  7 14:09 tomcat-juli.zip
